File: test/ModularPipelines.UnitTests/Engine/DependencyTreeFormatterTests.cs
using ModularPipelines.Context;
using ModularPipelines.Engine;
using ModularPipelines.Models;
using ModularPipelines.Modules;
using Spectre.Console;

namespace ModularPipelines.UnitTests.Engine;

/// <summary>
/// Tests for DependencyTreeFormatter Spectre.Console tree generation.
/// </summary>
public class DependencyTreeFormatterTests
{
    #region Test Modules

    private class ModuleA : Module<bool>
    {
        protected internal override Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
            => Task.FromResult(true);
    }

    private class ModuleB : Module<bool>
    {
        protected internal override Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
            => Task.FromResult(true);
    }

    private class ModuleC : Module<bool>
    {
        protected internal override Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
            => Task.FromResult(true);
    }

    private class ModuleD : Module<bool>
    {
        protected internal override Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
            => Task.FromResult(true);
    }

    private class SharedModule : Module<bool>
    {
        protected internal override Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
            => Task.FromResult(true);
    }

    #endregion

    #region Helper Methods

    private static string RenderToString(Tree tree)
    {
        using var writer = new StringWriter();
        var console = AnsiConsole.Create(new AnsiConsoleSettings
        {
            Out = new AnsiConsoleOutput(writer),
            Ansi = AnsiSupport.No, // Disable ANSI for easier text assertions
        });
        console.Write(tree);
        return writer.ToString();
    }

    private static ModuleDependencyModel CreateModel<T>() where T : IModule, new()
    {
        return new ModuleDependencyModel(new T());
    }

    #endregion

    [Test]
    public async Task FormatTree_SingleModule_NoDependencies_ReturnsTreeWithSingleNode()
    {
        // Arrange
        var formatter = new DependencyTreeFormatter();
        var moduleA = CreateModel<ModuleA>();
        var roots = new[] { moduleA };

        // Act
        var tree = formatter.FormatTree(roots);
        var output = RenderToString(tree);

        // Assert
        await Assert.That(output).Contains("Module Dependencies");
        await Assert.That(output).Contains("ModuleA");
    }

    [Test]
    public async Task FormatTree_LinearChain_ReturnsTreeWithCorrectHierarchy()
    {
        // Arrange: A -> B -> C (A is root, B depends on A, C depends on B)
        var formatter = new DependencyTreeFormatter();
        var moduleA = CreateModel<ModuleA>();
        var moduleB = CreateModel<ModuleB>();
        var moduleC = CreateModel<ModuleC>();

        // Set up dependency chain: A is dependency for B, B is dependency for C
        moduleA.IsDependencyFor.Add(moduleB);
        moduleB.IsDependencyFor.Add(moduleC);

        var roots = new[] { moduleA };

        // Act
        var tree = formatter.FormatTree(roots);
        var output = RenderToString(tree);

        // Assert
        await Assert.That(output).Contains("ModuleA");
        await Assert.That(output).Contains("ModuleB");
        await Assert.That(output).Contains("ModuleC");
    }

    [Test]
    public async Task FormatTree_MultipleRoots_ReturnsTreeWithAllRoots()
    {
        // Arrange: Two independent modules
        var formatter = new DependencyTreeFormatter();
        var moduleA = CreateModel<ModuleA>();
        var moduleB = CreateModel<ModuleB>();

        var roots = new[] { moduleA, moduleB };

        // Act
        var tree = formatter.FormatTree(roots);
        var output = RenderToString(tree);

        // Assert
        await Assert.That(output).Contains("ModuleA");
        await Assert.That(output).Contains("ModuleB");
    }

    [Test]
    public async Task FormatTree_SharedModule_MarkedAsReference_OnSecondOccurrence()
    {
        // Arrange: A -> Shared, B -> Shared (diamond pattern)
        var formatter = new DependencyTreeFormatter();
        var moduleA = CreateModel<ModuleA>();
        var moduleB = CreateModel<ModuleB>();
        var shared = CreateModel<SharedModule>();

        moduleA.IsDependencyFor.Add(shared);
        moduleB.IsDependencyFor.Add(shared);

        var roots = new[] { moduleA, moduleB };

        // Act
        var tree = formatter.FormatTree(roots);
        var output = RenderToString(tree);

        // Assert
        await Assert.That(output).Contains("SharedModule");
        // Should have reference marker on second occurrence
        await Assert.That(output).Contains("(↑)");
    }

    [Test]
    public async Task FormatTree_DiamondDependency_ShowsReferenceMarkerForSharedLeaf()
    {
        // Arrange: A -> B -> D, A -> C -> D (diamond)
        var formatter = new DependencyTreeFormatter();
        var moduleA = CreateModel<ModuleA>();
        var moduleB = CreateModel<ModuleB>();
        var moduleC = CreateModel<ModuleC>();
        var moduleD = CreateModel<ModuleD>();

        moduleA.IsDependencyFor.Add(moduleB);
        moduleA.IsDependencyFor.Add(moduleC);
        moduleB.IsDependencyFor.Add(moduleD);
        moduleC.IsDependencyFor.Add(moduleD);

        var roots = new[] { moduleA };

        // Act
        var tree = formatter.FormatTree(roots);
        var output = RenderToString(tree);

        // Assert
        await Assert.That(output).Contains("ModuleA");
        await Assert.That(output).Contains("ModuleB");
        await Assert.That(output).Contains("ModuleC");
        await Assert.That(output).Contains("ModuleD");
        // D appears twice, second should have reference marker
        await Assert.That(output).Contains("(↑)");
    }

    [Test]
    public async Task FormatTree_EmptyCollection_OnlyContainsHeader()
    {
        // Arrange
        var formatter = new DependencyTreeFormatter();
        var roots = Array.Empty<ModuleDependencyModel>();

        // Act
        var tree = formatter.FormatTree(roots);
        var output = RenderToString(tree);

        // Assert - should have header but no child nodes (no tree connectors)
        await Assert.That(output).Contains("Module Dependencies");
        await Assert.That(output).DoesNotContain("├");
        await Assert.That(output).DoesNotContain("└");
    }

    [Test]
    public async Task FormatTree_AlreadyPrintedRoot_SkipsIt()
    {
        // Arrange: Same module appearing as root twice shouldn't duplicate
        var formatter = new DependencyTreeFormatter();
        var moduleA = CreateModel<ModuleA>();

        // Add same root twice
        var roots = new[] { moduleA, moduleA };

        // Act
        var tree = formatter.FormatTree(roots);
        var output = RenderToString(tree);

        // Assert - should only appear once
        var count = output.Split("ModuleA").Length - 1;
        await Assert.That(count).IsEqualTo(1);
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Engine/DependencyTreeFormatterTests.fs
namespace ModularPipelines.UnitTests.FSharp.Engine

open System
open System.IO
open ModularPipelines.Context
open ModularPipelines.Engine
open ModularPipelines.Models
open ModularPipelines.Modules
open Spectre.Console
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type private ModuleA() =
    inherit Module<bool>()
    override _.ExecuteAsync(_: IModuleContext, _) = System.Threading.Tasks.Task.FromResult(true)

type private ModuleB() =
    inherit Module<bool>()
    override _.ExecuteAsync(_: IModuleContext, _) = System.Threading.Tasks.Task.FromResult(true)

type private ModuleC() =
    inherit Module<bool>()
    override _.ExecuteAsync(_: IModuleContext, _) = System.Threading.Tasks.Task.FromResult(true)

type private ModuleD() =
    inherit Module<bool>()
    override _.ExecuteAsync(_: IModuleContext, _) = System.Threading.Tasks.Task.FromResult(true)

type private SharedModule() =
    inherit Module<bool>()
    override _.ExecuteAsync(_: IModuleContext, _) = System.Threading.Tasks.Task.FromResult(true)

[<AutoOpen>]
module private DependencyTreeFormatterTestHelpers =
    let renderToString (tree: Tree) =
        use writer = new StringWriter()
        let settings = AnsiConsoleSettings()
        settings.Out <- AnsiConsoleOutput(writer)
        settings.Ansi <- AnsiSupport.No
        let console = AnsiConsole.Create(settings)
        console.Write(tree)
        writer.ToString()

    let createModel<'T when 'T :> IModule and 'T : (new: unit -> 'T)> () =
        ModuleDependencyModel(Activator.CreateInstance<'T>())

type DependencyTreeFormatterTests() =
    [<Test>]
    member _.FormatTree_SingleModule_NoDependencies_ReturnsTreeWithSingleNode() = async {
        let formatter = DependencyTreeFormatter()
        let moduleA = createModel<ModuleA>()
        let roots = [| moduleA |]

        let tree = formatter.FormatTree(roots)
        let output = renderToString tree

        do! check(Assert.That(output.Contains("Module Dependencies")).IsTrue())
        do! check(Assert.That(output.Contains("ModuleA")).IsTrue())
    }

    [<Test>]
    member _.FormatTree_LinearChain_ReturnsTreeWithCorrectHierarchy() = async {
        let formatter = DependencyTreeFormatter()
        let moduleA = createModel<ModuleA>()
        let moduleB = createModel<ModuleB>()
        let moduleC = createModel<ModuleC>()

        moduleA.IsDependencyFor.Add(moduleB)
        moduleB.IsDependencyFor.Add(moduleC)

        let tree = formatter.FormatTree([| moduleA |])
        let output = renderToString tree

        do! check(Assert.That(output.Contains("ModuleA")).IsTrue())
        do! check(Assert.That(output.Contains("ModuleB")).IsTrue())
        do! check(Assert.That(output.Contains("ModuleC")).IsTrue())
    }

    [<Test>]
    member _.FormatTree_MultipleRoots_ReturnsTreeWithAllRoots() = async {
        let formatter = DependencyTreeFormatter()
        let moduleA = createModel<ModuleA>()
        let moduleB = createModel<ModuleB>()

        let tree = formatter.FormatTree([| moduleA; moduleB |])
        let output = renderToString tree

        do! check(Assert.That(output.Contains("ModuleA")).IsTrue())
        do! check(Assert.That(output.Contains("ModuleB")).IsTrue())
    }

    [<Test>]
    member _.FormatTree_SharedModule_MarkedAsReference_OnSecondOccurrence() = async {
        let formatter = DependencyTreeFormatter()
        let moduleA = createModel<ModuleA>()
        let moduleB = createModel<ModuleB>()
        let shared = createModel<SharedModule>()

        moduleA.IsDependencyFor.Add(shared)
        moduleB.IsDependencyFor.Add(shared)

        let tree = formatter.FormatTree([| moduleA; moduleB |])
        let output = renderToString tree

        do! check(Assert.That(output.Contains("SharedModule")).IsTrue())
        do! check(Assert.That(output.Contains("(↑)")).IsTrue())
    }

    [<Test>]
    member _.FormatTree_DiamondDependency_ShowsReferenceMarkerForSharedLeaf() = async {
        let formatter = DependencyTreeFormatter()
        let moduleA = createModel<ModuleA>()
        let moduleB = createModel<ModuleB>()
        let moduleC = createModel<ModuleC>()
        let moduleD = createModel<ModuleD>()

        moduleA.IsDependencyFor.Add(moduleB)
        moduleA.IsDependencyFor.Add(moduleC)
        moduleB.IsDependencyFor.Add(moduleD)
        moduleC.IsDependencyFor.Add(moduleD)

        let tree = formatter.FormatTree([| moduleA |])
        let output = renderToString tree

        do! check(Assert.That(output.Contains("ModuleA")).IsTrue())
        do! check(Assert.That(output.Contains("ModuleB")).IsTrue())
        do! check(Assert.That(output.Contains("ModuleC")).IsTrue())
        do! check(Assert.That(output.Contains("ModuleD")).IsTrue())
        do! check(Assert.That(output.Contains("(↑)")).IsTrue())
    }

    [<Test>]
    member _.FormatTree_EmptyCollection_OnlyContainsHeader() = async {
        let formatter = DependencyTreeFormatter()
        let roots = Array.empty<ModuleDependencyModel>

        let tree = formatter.FormatTree(roots)
        let output = renderToString tree

        do! check(Assert.That(output.Contains("Module Dependencies")).IsTrue())
        do! check(Assert.That(output.Contains("├")).IsFalse())
        do! check(Assert.That(output.Contains("└")).IsFalse())
    }

    [<Test>]
    member _.FormatTree_AlreadyPrintedRoot_SkipsIt() = async {
        let formatter = DependencyTreeFormatter()
        let moduleA = createModel<ModuleA>()

        let tree = formatter.FormatTree([| moduleA; moduleA |])
        let output = renderToString tree
        let count = output.Split([| "ModuleA" |], StringSplitOptions.None).Length - 1
        
        do! check(IntEqualsAssertionExtensions.IsEqualTo(Assert.That(count), 1))
    }
